woocommerce/google-listings-and-ads
# Fix retries when updating products

### Changes proposed in this Pull Request:

Closes #1239 .

The task in the initial issue was to limit the amount of retries so if we would receive internal errors it wouldn't endlessly retry. However after additional testing it appears the `can_schedule` will check if there is a scheduled task running with the same hook name and the same list of arguments. When we call the action `woocommerce_gla_batch_retry_update_products` we are still running the scheduled task with the same hook name. So if all the products in the update batch fail we will have the same list of arguments and the retry will not be scheduled. This can occur easier if we are only updating one product at a time (for example when we set a single product to not sync or show, or it goes out of stock).

So this PR changes the following things:

- Always reschedule when we called the action `woocommerce_gla_batch_retry_update_products`
- Reschedule a retry after 1 minute (no use to retry an internal error right away)
- Limit retries to the product syncer threshold (5 tries)
- Add unit tests

### Screens:

<img width="1503" alt="WooCommerce_status_‹_Woo_s_👾_Commerce_—_WordPress" src="https://user-images.githubusercontent.com/9010963/159238912-43c13731-fb81-4fd7-883d-05a33345e3d0.png">

### Detailed test instructions:

Since we can't simulate internal errors in the Google API's we will need to test this PR with our test proxy in the `update_error` mode so it can return mocked error responses. If you are using **Docker see instructions** [here](https://github.com/woocommerce/google-listings-and-ads/blob/develop/tests/proxy/proxy.md#connect-test-site-to-proxy) on how to get it running.

- Run the proxy with a command like: ```PROXY_MODE=update_error cd tests/proxy && npm run test-proxy```
- Use a code snippet to get our site to connect to the proxy:
`define( 'WOOCOMMERCE_GLA_CONNECT_SERVER_URL', 'http://localhost:5500' );`
- Connect your Merchant account and sync all or some products (ensure we waited for the syncing to complete)
- Edit one of the synced products and mark it as "Sync and show" and save it
- This will trigger a product update, since we mocked the errors it will trigger a retry until it reaches the threshold of 5
- Wait for at least 6 minutes till it completes all the retries
- View the GLA log file in WooCommerce > Status > Logs
- Confirm that you see 4 failures in the log with a message like (the 5th time it won't be retried so it will only show debug messages that it failed):
```
2022-03-21T09:54:17+00:00 ERROR Automattic\WooCommerce\GoogleListingsAndAds\Product\ProductSyncer::handle_update_errors Internal API errors while submitting the following products: 33
```
- Check in WooCommerce > Status > Scheduled Actions and confirm that we have no more pending actions with the hook `gla/jobs/update_products/process_item`
- If we check the completed actions we will see 5 with the hook `gla/jobs/update_products/process_item`

### Changelog entry

- Fix - Limit failed update retries to 5 and schedule again after one minute.
